core/drivers/imx/dcp/dcp.c

```diff
@@ -0,0 +1,776 @@
+
+	/* Update descriptor structure to be processed for the channel */
+	io_write32(dcp_base + DCP_CH_N_CMDPTR(dcp_data->channel),
+		   virt_to_phys(&dcp_data->desc));
+
+	/* Increment the semaphore to start the transfer */
+	io_write32(dcp_base + DCP_CH_N_SEMA(dcp_data->channel), 0x1);
+
+	for (timeout = 0; timeout < DCP_MAX_TIMEOUT; timeout++) {
+		dcp_udelay(10);
+		val = io_read32(dcp_base + DCP_STAT);
+		if (val & BIT(dcp_data->channel))
+			break;
+	}
+
+	val = io_read32(dcp_base + DCP_CH_N_STAT(dcp_data->channel));
+	if (val & DCP_CH_STAT_ERROR_MASK) {
+		EMSG("Error operation, 0x%" PRIx32, val);
+		ret = TEE_ERROR_GENERIC;
+	}
+
+	if (timeout == DCP_MAX_TIMEOUT) {
+		EMSG("Timeout elapsed before operation");
+		ret = TEE_ERROR_GENERIC;
+	}
+
+	dcp_clk_enable(false);
+
+	return ret;
+}
+
+static TEE_Result dcp_cmac_subkey_generation(struct dcp_cipher_init *init,
+					     uint8_t *k1, uint8_t *k2)
+{
+	TEE_Result ret = TEE_ERROR_GENERIC;
+	struct dcp_cipher_data data = { };
+	uint8_t l[16] = { };
+	uint8_t tmp[16] = { };
+	uint8_t const_zero[16] = { };
+	uint8_t const_rb[16] = { [15] = 0x87 };
+
+	ret = dcp_cipher_do_init(&data, init);
+	if (ret != TEE_SUCCESS)
+		return ret;
+
+	ret = dcp_cipher_do_update(&data, const_zero, l, sizeof(l));
+	if (ret != TEE_SUCCESS)
+		goto out;
+
+	if ((l[0] & BIT(7)) == 0) {
+		dcp_left_shift_buffer(l, k1, 16);
+	} else {
+		dcp_left_shift_buffer(l, tmp, 16);
+		dcp_xor(tmp, const_rb, k1, 16);
+	}
+
+	if ((k1[0] & BIT(7)) == 0) {
+		dcp_left_shift_buffer(k1, k2, 16);
+	} else {
+		dcp_left_shift_buffer(k1, tmp, 16);
+		dcp_xor(tmp, const_rb, k2, 16);
+	}
+
+	ret = TEE_SUCCESS;
+out:
+	dcp_cipher_do_final(&data);
+
+	return ret;
+}
+
+TEE_Result dcp_store_key(uint32_t *key, unsigned int index)
+{
+	uint32_t val = 0;
+	unsigned int i = 0;
+	uint32_t key_store_except = 0;
+
+	if (!key)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (index > DCP_SRAM_KEY_NB_SUBWORD - 1) {
+		EMSG("Bad parameters, index must be < %u",
+		     DCP_SRAM_KEY_NB_SUBWORD);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	key_store_except = cpu_spin_lock_xsave(&key_store_spinlock);
+
+	dcp_clk_enable(true);
+
+	val = DCP_SRAM_KEY_INDEX(index);
+	io_write32(dcp_base + DCP_KEY, val);
+
+	/*
+	 * Key is stored as four uint32 values, starting with subword0
+	 * (least-significant word)
+	 */
+	for (i = 0; i < DCP_SRAM_KEY_NB_SUBWORD; i++) {
+		val = TEE_U32_TO_BIG_ENDIAN(key[i]);
+		io_write32(dcp_base + DCP_KEYDATA, val);
+	}
+
+	dcp_clk_enable(false);
+
+	cpu_spin_unlock_xrestore(&key_store_spinlock, key_store_except);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result dcp_cmac(struct dcp_cipher_init *init, uint8_t *input,
+		    size_t input_size, uint8_t *output)
+{
+	TEE_Result ret = TEE_ERROR_GENERIC;
+	uint8_t key1[DCP_AES128_KEY_SIZE] = { };
+	uint8_t key2[DCP_AES128_KEY_SIZE] = { };
+	unsigned int nb_blocks = 0;
+	bool block_complete = false;
+	struct dcp_cipher_data data = { };
+	struct dcp_align_buf x = { };
```

Comment:
Why do `x`, `y` and `last` need to be of `struct dcp_align_buf`. It seems that `dcp_cipher_do_update()` always allocates temporary buffers anyway.

Reply:
You're right. I will use buffers ```uint8_t x[DCP_AES128_KEY_SIZE] = { };```